libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/layoutselection/OpenHouseDataLayoutGenerator.java

```diff
@@ -0,0 +1,121 @@
+  private static final long NUM_FILE_GROUPS_PER_COMMIT = 50;
+  private static final double FILE_ENTROPY_THRESHOLD = 100.0;
+  private static final long FILE_OPEN_COST_IN_BYTES = 1024L * 1024L * 4; // 4MB
+  private static final long REWRITE_BYTES_PER_SECOND = 1024L * 1024L * 2; // 2MB/s
+  private static final int REWRITE_PARALLELISM = 900; // number of Spark tasks to run in parallel
+  private final TableFileStats tableFileStats;
+
+  /**
+   * Generate a list of data layout optimization strategies based on the table file stats and
+   * historic query patterns.
+   */
+  @Override
+  public List<DataLayoutOptimizationStrategy> generate() {
+    return Collections.singletonList(generateCompactionStrategy());
+  }
+
+  private DataLayoutOptimizationStrategy generateCompactionStrategy() {
+    long totalSizeBytes =
+        tableFileStats
+            .get()
+            .map((MapFunction<FileStat, Long>) FileStat::getSize, Encoders.LONG())
+            .reduce((ReduceFunction<Long>) Long::sum);
+
+    DataCompactionConfig.DataCompactionConfigBuilder configBuilder = DataCompactionConfig.builder();
+
+    long targetBytesSize = calculateTargetBytesSize();
+    configBuilder.targetByteSize(targetBytesSize);
+
+    // TODO: take partitioning into account
+    // This is a under-estimation if partitions are smaller than
+    // DataCompactionConfig.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT
+    int estimatedMaxNumFileGroups =
+        Math.max(
+            1, (int) (totalSizeBytes / DataCompactionConfig.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT));
+
+    int maxNumCommits =
+        (int)
+            Math.min(
+                1000L,
+                (estimatedMaxNumFileGroups + NUM_FILE_GROUPS_PER_COMMIT - 1)
+                    / NUM_FILE_GROUPS_PER_COMMIT);
+    configBuilder.partialProgressMaxCommits(maxNumCommits);
+
+    int estimatedNumTasksPerFileGroup =
+        (int) (DataCompactionConfig.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT / targetBytesSize);
+
+    // TODO: take partitioning into account
+    // This is a under-estimation if partitions are much smaller than
+    // DataCompactionConfig.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT
+    configBuilder.maxConcurrentFileGroupRewrites(
+        Math.min(
+            50,
+            (estimatedNumTasksPerFileGroup + REWRITE_PARALLELISM - 1)
+                / estimatedNumTasksPerFileGroup));
+
+    // don't split large files
+    configBuilder.maxByteSizeRatio(10);
+
+    return DataLayoutOptimizationStrategy.builder()
+        .config(configBuilder.build())
+        .cost(calculateCompactionCost())
+        .gain(calculateCompactionGain())
+        .build();
+  }
+
+  private long calculateTargetBytesSize() {
+    // Make sure the last block is almost full
+    return 2 * FILE_BLOCK_SIZE_BYTES - FILE_BLOCK_MARGIN_BYTES;
+  }
+
+  /** Calculate the gain of compaction in terms of number of files reduced. */
+  private double calculateCompactionGain() {
```

Comment:
Is this the right layering to have computing gain and cost as part of OpenHouseDataLayoutGenerator? I thought OpenHouseDataLayoutGenerator will generate strategy. Then there is a separate module that executes the optimization strategy, which should compute gain/score and stamp it with the respective strategy in table properties.

Reply:
It's a bit variable, but let me add README on how it's currently designed to work